File: working dir/widgets/playlistView/PlaylistView.java
package playlistView;

import java.util.Collection;
import java.util.List;
import java.util.UUID;
import javafx.event.Event;
import javafx.fxml.FXML;
import javafx.geometry.NodeOrientation;
import javafx.scene.control.Menu;
import javafx.scene.control.Tooltip;
import javafx.scene.layout.AnchorPane;
import sp.it.pl.audio.Item;
import sp.it.pl.audio.Player;
import sp.it.pl.audio.playlist.Playlist;
import sp.it.pl.audio.playlist.PlaylistItem;
import sp.it.pl.audio.playlist.PlaylistItem.Field;
import sp.it.pl.audio.playlist.PlaylistManager;
import sp.it.pl.gui.Gui;
import sp.it.pl.gui.objects.icon.Icon;
import sp.it.pl.gui.objects.table.PlaylistTable;
import sp.it.pl.gui.objects.table.TableColumnInfo;
import sp.it.pl.layout.widget.Widget;
import sp.it.pl.layout.widget.controller.FXMLController;
import sp.it.pl.layout.widget.controller.io.Output;
import sp.it.pl.layout.widget.feature.PlaylistFeature;
import sp.it.pl.layout.widget.feature.SongReader;
import sp.it.pl.util.access.V;
import sp.it.pl.util.access.Vo;
import sp.it.pl.util.async.executor.ExecuteN;
import sp.it.pl.util.conf.Config;
import sp.it.pl.util.conf.IsConfig;
import sp.it.pl.util.type.Util;
import sp.it.pl.util.units.Dur;
import static de.jensd.fx.glyphs.fontawesome.FontAwesomeIcon.FILTER;
import static de.jensd.fx.glyphs.materialdesignicons.MaterialDesignIcon.FILTER_OUTLINE;
import static java.util.stream.Collectors.toList;
import static javafx.scene.control.SelectionMode.MULTIPLE;
import static sp.it.pl.gui.infonode.InfoTable.DEFAULT_TEXT_FACTORY;
import static sp.it.pl.layout.widget.Widget.Group.PLAYLIST;
import static sp.it.pl.layout.widget.WidgetManager.WidgetSource.NO_LAYOUT;
import static sp.it.pl.layout.widget.WidgetManager.WidgetSource.OPEN;
import static sp.it.pl.main.App.APP;
import static sp.it.pl.main.AppBuildersKt.appTooltip;
import static sp.it.pl.util.functional.Util.ISNTØ;
import static sp.it.pl.util.functional.Util.list;
import static sp.it.pl.util.functional.Util.stream;
import static sp.it.pl.util.graphics.Util.menuItem;
import static sp.it.pl.util.graphics.Util.setAnchors;
import static sp.it.pl.util.reactive.Util.maintain;

/**
 * Playlist FXML Controller class
 * Controls behavior of the Playlist FXML graphics.
 */
@Widget.Info(
    author = "Martin Polakovic",
    name = "Playlist",
    description = "Provides list of items to play. Highlights playing and unplayable "
                + "items.",
    howto = "Available actions:\n" +
            "    Item left click : Selects item\n" +
            "    Item right click : Opens context menu\n" +
            "    Item double click : Plays item\n" +
            "    Item drag : Activates Drag&Drop\n" +
            "    Item drag + CTRL : Moves item within playlist\n" +
            "    Type : search & filter\n" +
            "    Press ENTER : Plays item\n" +
            "    Press ESC : Clear selection & filter\n" +
            "    Scroll : Scroll table vertically\n" +
            "    Scroll + SHIFT : Scroll table horizontally\n" +
            "    Column drag : swap columns\n" +
            "    Column right click: show column menu\n" +
            "    Click column : Sort - ascending | descending | none\n" +
            "    Click column + SHIFT : Sorts by multiple columns\n" +
            "    Menu bar : Opens additional actions\n",
    notes = "Plans: multiple playlists through tabs",
    version = "1",
    year = "2015",
    group = PLAYLIST
)
public class PlaylistView extends FXMLController implements PlaylistFeature {

    private @FXML AnchorPane root;
    private Playlist playlist;
    private PlaylistTable table;
    private final ExecuteN columnInitializer = new ExecuteN(1);

    private Output<PlaylistItem> outSelected, outPlaying;

    // configurables
    @IsConfig(name = "Table orientation", info = "Orientation of the table.")
    public final Vo<NodeOrientation> orient = new Vo<>(Gui.table_orient);
    @IsConfig(name = "Zeropad numbers", info = "Adds 0s for number length consistency.")
    public final Vo<Boolean> zeropad = new Vo<>(Gui.table_zeropad);
    @IsConfig(name = "Search show original index", info = "Show unfiltered table item index when filter applied.")
    public final Vo<Boolean> orig_index = new Vo<>(Gui.table_orig_index);
    @IsConfig(name = "Show table header", info = "Show table header with columns.")
    public final Vo<Boolean> show_header = new Vo<>(Gui.table_show_header);
    @IsConfig(name = "Show table footer", info = "Show table controls at the bottom of the table. Displays menubar and table items information.")
    public final Vo<Boolean> show_footer = new Vo<>(Gui.table_show_footer);
    @IsConfig(name = "Scroll to playing", info = "Scroll table to playing item when it changes.")
    public final V<Boolean> scrollToPlaying = new V<>(true);
    @IsConfig(name = "Play displayed only", info = "Only displayed items will be played when filter is active.")
    public final V<Boolean> filter_for_playback = new V<>(false, v -> {
        String of = "Enable filter for playback. Causes the playback "
                  + "to play only displayed items.";
        String on = "Disable filter for playback. Causes the playback "
                  + "to ignore the filter.";
        Tooltip t = appTooltip(v ? on : of);
                t.setWrapText(true);
                t.setMaxWidth(200);
        Icon i = table.filterPane.getButton();
             i.icon(v ? FILTER : FILTER_OUTLINE);
             i.onClick(this::filterToggle);
             i.tooltip(v ? on : of);
             i.setDisable(false); // needed
        setUseFilterForPlayback(v);
    });

    @Override
    public void init() {

        // obtain playlist by id, we will use this widget's id
        UUID id = getWidget().id;
        playlist = PlaylistManager.playlists.computeIfAbsent(id, PlaylistView::getUnusedPlaylist);
        // when widget closes we must remove the playlist or it would get saved
        // and playlist list would infinitely grow. When widgets close naturally
        // on app close, the playlist will get removed after app state was saved => no problem
        d(() -> PlaylistManager.playlists.remove(playlist));

        // widget input/output
        outSelected = outputs.create(id,"Selected", Item.class, null);
        outPlaying = outputs.create(id,"Playing", Item.class, null);
        d(Player.playlistSelected.i.bind(outSelected));
        d(maintain(playlist.playingI, ι -> playlist.getPlaying(), outPlaying));
        d(Player.onItemRefresh(ms -> {
            if (outPlaying.getValue()!=null)
                ms.ifHasK(outPlaying.getValue().getUri(), m -> outPlaying.setValue(m.toPlaylist()));
            if (outSelected.getValue()!=null)
                ms.ifHasK(outSelected.getValue().getUri(), m -> outSelected.setValue(m.toPlaylist()));
        }));

        table = new PlaylistTable(playlist);

        // add table to scene graph
        root.getChildren().add(table.getRoot());
        setAnchors(table.getRoot(),0d);

        // table properties
        table.setFixedCellSize(Gui.font.getValue().getSize() + 5);
        table.getSelectionModel().setSelectionMode(MULTIPLE);
        d(maintain(orient,table.nodeOrientationProperty()));
        d(maintain(zeropad,table.zeropadIndex));
        d(maintain(orig_index,table.showOriginalIndex));
        d(maintain(show_header,table.headerVisible));
        d(maintain(show_footer,table.footerVisible));
        d(maintain(scrollToPlaying,table.scrollToPlaying));

        // extend table items information
        table.items_info.textFactory = (all, list) -> {
            double Σms = list.stream().mapToDouble(PlaylistItem::getTimeMs).sum();
            return DEFAULT_TEXT_FACTORY.apply(all, list) + " - " + new Dur(Σms);
        };
        // add more menu items
        table.menuAdd.getItems().addAll(
            menuItem("Add files", e -> PlaylistManager.chooseFilesToAdd()),
            menuItem("Add directory", e -> PlaylistManager.chooseFolderToAdd()),
            menuItem("Add URL", e -> PlaylistManager.chooseUrlToAdd()),
            menuItem("Play files", e -> PlaylistManager.chooseFilesToPlay()),
            menuItem("Play directory", e -> PlaylistManager.chooseFolderToPlay()),
            menuItem("Play URL", e -> PlaylistManager.chooseUrlToPlay()),
            menuItem("Duplicate selected (+)", e -> playlist.duplicateItemsByOne(table.getSelectedItems())),
            menuItem("Duplicate selected (*)", e -> playlist.duplicateItemsAsGroup(table.getSelectedItems()))
//            editOnAdd_menuItem    // add to lib option
//            editOnAdd_menuItem    // add to lib + edit option
        );
        table.menuRemove.getItems().addAll(
            menuItem("Remove selected", e -> playlist.removeAll(table.getSelectedItems())),
            menuItem("Remove not selected", e -> playlist.retainAll(table.getSelectedItems())),
            menuItem("Remove unsupported", e -> playlist.removeUnplayable()),
            menuItem("Remove duplicates", e -> playlist.removeDuplicates()),
            menuItem("Remove all", e -> playlist.clear())
        );
        table.menuOrder.getItems().addAll(
            menuItem("Order reverse", e -> playlist.reverse()),
            menuItem("Order randomly", e -> playlist.randomize()),
            menuItem("Edit selected", e -> APP.widgetManager.use(SongReader.class,NO_LAYOUT,w->w.read(table.getSelectedItems())))
            // menuItem("Save selected as...", e -> saveSelectedAsPlaylist())
        );
        Menu sortM = new Menu("Order by");
        stream(Field.FIELDS)
                .map(f -> menuItem(f.toStringEnum(), e -> table.sortBy(f)))
                .forEach(sortM.getItems()::add);
        table.menuOrder.getItems().add(0, sortM);

        // prevent scroll event to propagate up
        root.setOnScroll(Event::consume);

        // maintain outputs
        table.getSelectionModel().selectedItemProperty().addListener((o,ov,nv) -> {
            if (!table.movingItems)
                outSelected.setValue(nv);
        });

        d(table::dispose);
    }

    @Override
    public Collection<Config<Object>> getFields() {
        // serialize column state when requested
        getWidget().properties.put("columns", table.getColumnState().toString());
        return super.getFields();
    }

    @Override
    public void refresh() {
        columnInitializer.execute(() -> {
            String c = getWidget().properties.getS("columns");
            table.setColumnState(c==null ? table.getDefaultColumnInfo() : TableColumnInfo.fromString(c));
        });
        filter_for_playback.applyValue();
    }

    @Override
    public Playlist getPlaylist() {
        return playlist;
    }

    // TODO: implement properly
//    void saveSelectedAsPlaylist() {
//        List<PlaylistItem> l = table.getSelectedItems();
//        if (l.isEmpty()) return;
//
//        App.APP.actions.doWithUserString("Save selected items as...", "Name",
//        Config<String> mc = new ValueConfig<>(String.class, "Name", "My Playlist").constraints(new StringNonEmpty());
//        SimpleConfigurator<?> sc = new SimpleConfigurator<>(mc, (String name) -> {
//            Playlist p = new Playlist(UUID.randomUUID());
//                     p.setAll(l);
//            APP.serializators.toXML(p, new File(APP.DIR_PLAYLISTS, name + ".xml"))
//                .ifError(e -> log(PlaylistView.class).error("Could not save playlist", e));
//        });
//        PopOver p = new PopOver<>(sc);
//                p.title.set("Save selected items as...");
//                p.show(ScreenPos.APP_CENTER);
//    }

    private void setUseFilterForPlayback(boolean v) {
        playlist.setTransformation(v
            ? orig -> list(table.getItems())
            : orig -> orig.stream().sorted(table.itemsComparator.get()).collect(toList())
        );
    }

    private void filterToggle() {
        filter_for_playback.setCycledNapplyValue();
    }

    private static Playlist getUnusedPlaylist(UUID id) {
        List<Playlist> pall = list(PlaylistManager.playlists);
        APP.widgetManager.findAll(OPEN).filter(w -> w.getInfo().hasFeature(PlaylistFeature.class))
                 .filter(w -> w.getController()!=null) // during load some widgets may not be loaded yet, not good
                 .map(w -> ((PlaylistFeature)w.getController()).getPlaylist())
                 .filter(ISNTØ)
                 .forEach(p -> pall.removeIf(pl -> pl.id.equals(p.id)));

        Playlist leaf = pall.isEmpty() ? null : pall.get(0);
        for (Playlist p : pall)
            if (p.id.equals(PlaylistManager.active))
                leaf = p;
        if (leaf!=null) PlaylistManager.playlists.remove(leaf);
        if (leaf!=null) {
            if (leaf.id.equals(PlaylistManager.active))
                PlaylistManager.active = id;
            Util.setField(Playlist.class, leaf, "id", id);
        }
        return leaf==null ? new Playlist(id) : leaf;
    }
}

File: working dir/widgets/playlistView/PlaylistView.kt
package playlistView

import javafx.event.EventHandler
import javafx.geometry.NodeOrientation.INHERIT
import javafx.scene.control.SelectionMode.MULTIPLE
import javafx.stage.FileChooser
import sp.it.pl.audio.Player
import sp.it.pl.audio.PlayerConfiguration
import sp.it.pl.audio.playlist.Playlist
import sp.it.pl.audio.playlist.PlaylistItem
import sp.it.pl.audio.playlist.PlaylistItem.Field
import sp.it.pl.audio.playlist.PlaylistManager
import sp.it.pl.audio.playlist.writePlaylist
import sp.it.pl.gui.nodeinfo.TableInfo.Companion.DEFAULT_TEXT_FACTORY
import sp.it.pl.gui.objects.table.PlaylistTable
import sp.it.pl.gui.objects.table.TableColumnInfo
import sp.it.pl.layout.widget.Widget
import sp.it.pl.layout.widget.Widget.Group
import sp.it.pl.layout.widget.WidgetSource.NO_LAYOUT
import sp.it.pl.layout.widget.WidgetSource.OPEN
import sp.it.pl.layout.widget.controller.SimpleController
import sp.it.pl.layout.widget.controller.io.Output
import sp.it.pl.layout.widget.feature.PlaylistFeature
import sp.it.pl.layout.widget.feature.SongReader
import sp.it.pl.main.APP
import sp.it.pl.main.IconFA
import sp.it.pl.main.IconMD
import sp.it.pl.main.Widgets.PLAYLIST
import sp.it.pl.util.access.Vo
import sp.it.pl.util.async.runNew
import sp.it.pl.util.collections.materialize
import sp.it.pl.util.conf.Config
import sp.it.pl.util.conf.IsConfig
import sp.it.pl.util.conf.cn
import sp.it.pl.util.conf.cv
import sp.it.pl.util.conf.only
import sp.it.pl.util.file.parentDirOrRoot
import sp.it.pl.util.functional.net
import sp.it.pl.util.functional.orNull
import sp.it.pl.util.graphics.item
import sp.it.pl.util.graphics.items
import sp.it.pl.util.graphics.layFullArea
import sp.it.pl.util.graphics.menu
import sp.it.pl.util.reactive.attach
import sp.it.pl.util.reactive.on
import sp.it.pl.util.reactive.sync
import sp.it.pl.util.reactive.syncFrom
import sp.it.pl.util.system.saveFile
import sp.it.pl.util.type.Util
import sp.it.pl.util.units.Dur
import sp.it.pl.util.validation.Constraint.FileActor.DIRECTORY
import java.util.UUID
import java.util.function.Consumer
import java.util.function.UnaryOperator
import kotlin.streams.asSequence

@Widget.Info(
        author = "Martin Polakovic",
        name = PLAYLIST,
        description = "Provides list of items to play. Highlights playing and unplayable "+"items.",
        howto = ""
                +"Available actions:\n"
                +"    Item left click : Selects item\n"
                +"    Item right click : Opens context menu\n"
                +"    Item double click : Plays item\n"
                +"    Item drag : Activates Drag&Drop\n"
                +"    Item drag + CTRL : Moves item within playlist\n"
                +"    Type : search & filter\n"
                +"    Press ENTER : Plays item\n"
                +"    Press ESC : Clear selection & filter\n"
                +"    Scroll : Scroll table vertically\n"
                +"    Scroll + SHIFT : Scroll table horizontally\n"
                +"    Column drag : swap columns\n"
                +"    Column right click: show column menu\n"
                +"    Click column : Sort - ascending | descending | none\n"
                +"    Click column + SHIFT : Sorts by multiple columns\n"
                +"    Menu bar : Opens additional actions\n",
        notes = "Plans: multiple playlists through tabs",
        version = "1",
        year = "2015",
        group = Group.PLAYLIST
)
class PlaylistView(widget: Widget<*>): SimpleController(widget), PlaylistFeature {

    private val playlist = computeInitialPlaylist(widget.id)
    private val table = PlaylistTable(playlist)
    private var outSelected: Output<PlaylistItem?> = outputs.create(widget.id, "Selected", PlaylistItem::class.java, null)
    private var outPlaying: Output<PlaylistItem?> = outputs.create(widget.id, "Playing", PlaylistItem::class.java, null)

    @IsConfig(name = "Table orientation", info = "Orientation of the table.")
    val tableOrient by cv(INHERIT) { Vo(APP.ui.tableOrient) }
    @IsConfig(name = "Zeropad numbers", info = "Adds 0s for number length consistency.")
    val tableZeropad by cv(true) { Vo(APP.ui.tableZeropad) }
    @IsConfig(name = "Search show original index", info = "Show unfiltered table item index when filter applied.")
    val tableOrigIndex by cv(true) { Vo(APP.ui.tableOrigIndex) }
    @IsConfig(name = "Show table header", info = "Show table header with columns.")
    val tableShowHeader by cv(true) { Vo(APP.ui.tableShowHeader) }
    @IsConfig(name = "Show table footer", info = "Show table controls at the bottom of the table. Displays menubar and table items information.")
    val tableShowFooter by cv(true) { Vo(APP.ui.tableShowFooter) }
    @IsConfig(name = "Scroll to playing", info = "Scroll table to playing item when it changes.")
    val scrollToPlaying by cv(true)
    @IsConfig(name = "Play displayed only", info = "Only displayed items will be played when filter is active.")
    val playVisible by cv(false)
    @IsConfig(name = "Default browse location", info = "Opens this location for file dialogs.")
    var lastSavePlaylistLocation by cn(APP.DIR_USERDATA).only(DIRECTORY)

    init {
        playlist.playingI sync { outPlaying.value = playlist.playing } on onClose
        Player.playlistSelected.i.bind(outSelected) on onClose
        Player.onItemRefresh { ms ->
            outPlaying.value?.let { ms.ifHasK(it.uri, Consumer { outPlaying.value = it.toPlaylist() }) }
            outSelected.value?.let { ms.ifHasK(it.uri, Consumer { outSelected.value = it.toPlaylist() }) }
        } on onClose

        playVisible sync {
            table.filterPane.button.icon(if (it) IconFA.FILTER else IconMD.FILTER_OUTLINE)
            table.filterPane.button.onClickDo { playVisible.setCycledValue() }
            table.filterPane.button.tooltip(
                    if (it) "Disable filter for playback. Causes the playback "+"to ignore the filter."
                    else "Enable filter for playback. Causes the playback "+"to play only displayed items."
            )
            table.filterPane.button.isDisable = false // needed
            playlist.setTransformation(
                    if (it) unOp { table.items.materialize() }
                    else unOp { it.asSequence().sortedWith(table.itemsComparator.value).toList() }
            )
        } on onClose

        table.search.setColumn(Field.NAME)
        table.selectionModel.selectionMode = MULTIPLE
        table.items_info.textFactory = { all, list -> DEFAULT_TEXT_FACTORY(all, list)+" - "+Dur(list.sumByDouble { it.timeMs }) }
        table.nodeOrientationProperty() syncFrom tableOrient on onClose
        table.zeropadIndex syncFrom tableZeropad on onClose
        table.showOriginalIndex syncFrom tableOrigIndex on onClose
        table.headerVisible syncFrom tableShowHeader on onClose
        table.footerVisible syncFrom tableShowFooter on onClose
        table.scrollToPlaying syncFrom scrollToPlaying on onClose
        table.columnState = widget.properties.getS("columns")?.net { TableColumnInfo.fromString(it) } ?: table.defaultColumnInfo

        layFullArea += table.root

        table.menuAdd.apply {
            item("Add files") { PlaylistManager.chooseFilesToAdd() }
            item("Add directory") { PlaylistManager.chooseFolderToAdd() }
            item("Add URL") { PlaylistManager.chooseUrlToAdd() }
            item("Play files") { PlaylistManager.chooseFilesToPlay() }
            item("Play directory") { PlaylistManager.chooseFolderToPlay() }
            item("Play URL") { PlaylistManager.chooseUrlToPlay() }
            item("Duplicate selected (+)") { playlist.duplicateItemsByOne(table.selectedItems) }
            item("Duplicate selected (*)") { playlist.duplicateItemsAsGroup(table.selectedItems) }
        }
        table.menuRemove.apply {
            item("Remove selected") { playlist -= table.selectedItems }
            item("Remove not selected") { playlist.retainAll(table.selectedItems) }
            item("Remove unplayable") { playlist.removeUnplayable() }
            item("Remove duplicates") { playlist.removeDuplicates() }
            item("Remove all") { playlist.clear() }
        }
        table.menuOrder.apply {
            menu("Order by") {
                items(Field.FIELDS.asSequence(), { it.toStringEnum() }) { table.sortBy(it) }
            }
            item("Order reverse") { playlist.reverse() }
            item("Order randomly") { playlist.randomize() }
            item("Edit selected") { APP.widgetManager.widgets.use<SongReader>(NO_LAYOUT) { it.read(table.selectedItems) } }
            item("Save playlist") {
                saveFile(
                        "Save playlist as...",
                        lastSavePlaylistLocation ?: PlayerConfiguration.lastSavePlaylistLocation,
                        "Playlist",
                        widget.windowOrActive.orNull()?.stage ?: APP.windowManager.createStageOwner(),
                        FileChooser.ExtensionFilter("m3u8", "m3u8")
                ).ifOk { file ->
                    lastSavePlaylistLocation = file.parentDirOrRoot
                    PlayerConfiguration.lastSavePlaylistLocation = file.parentDirOrRoot
                    runNew { writePlaylist(table.selectedOrAllItemsCopy, file.name, file.parentDirOrRoot) }
                }
            }
        }

        onClose += table.selectionModel.selectedItemProperty() attach {
            if (!table.movingItems)
                outSelected.value = it
        }
        onClose += table::dispose

        onScroll = EventHandler { it.consume() }
    }

    override fun getFields(): Collection<Config<Any>> {
        widget.properties["columns"] = table.columnState.toString()
        return super.getFields()
    }

    override fun getPlaylist() = playlist

    private fun computeInitialPlaylist(id: UUID) = null
            ?: PlaylistManager.playlists[id]
            ?: getUnusedPlaylist(id).apply {
                PlaylistManager.playlists.put(this)
            }

    companion object {

        fun unOp(block: (List<PlaylistItem>) -> List<PlaylistItem>) = UnaryOperator<List<PlaylistItem>> { block(it) }

        infix fun <T> MutableList<T>.addToStart(item: T) = add(0, item)

        private fun getUnusedPlaylist(id: UUID): Playlist {
            val danglingPlaylists = ArrayList(PlaylistManager.playlists)
            APP.widgetManager.widgets.findAll(OPEN).asSequence()
                    .filter { it.info.hasFeature(PlaylistFeature::class.java) }
                    .mapNotNull { (it.controller as PlaylistFeature?)?.playlist }
                    .forEach { danglingPlaylists.removeIf { playlist -> playlist.id==it.id } }

            val danglingPlaylist: Playlist? = null
                    ?: danglingPlaylists.find { it.id==PlaylistManager.active }
                    ?: danglingPlaylists.firstOrNull()

            if (danglingPlaylist!=null) {
                PlaylistManager.playlists.remove(danglingPlaylist)

                if (danglingPlaylist.id==PlaylistManager.active)
                    PlaylistManager.active = id
                Util.setField(Playlist::class.java, danglingPlaylist, Playlist::id.name, id) // TODO: fix
            }

            return danglingPlaylist ?: Playlist(id)
        }

    }

}